README.md
# P2PFile-Sharing
# MiniDropBox
# MiniDropBox


201501113/history
2017-03-09 21:37:57	IndexGet shortlist
2017-03-09 21:40:34	IndexGet longlist
2017-03-09 21:42:42	IndexGet longlist
2017-03-13 17:12:43	index longlist
2017-03-13 17:13:58	index shortlist
2017-03-13 17:16:34	index regex ^t
2017-03-13 17:17:34	index regex ^t
2017-03-13 17:18:03	download server.pyc
2017-03-13 17:18:24	download test2.pyc
2017-03-13 17:18:40	download test2.txt
2017-03-13 17:19:19	hash
2017-03-13 17:19:43	hash get test2.txt
2017-03-13 17:20:37	index regex \.txt$
2017-03-13 17:20:50	hash checkall
2017-03-13 17:21:16	hash verify test2.txt
2017-03-13 17:23:54	index longlist ..
2017-03-13 19:57:19	index shortlist Mon Mar 13 19:56:50 2017 Mon Mar 13 19:56:45 2017
2017-03-13 19:58:43	index shortlist Mon Mar 13 19:50:50 2017 Mon Mar 13 19:56:45 2017
2017-03-13 20:10:24	index shortlist ?Mon Mar 13 20:08:35 2017 ? Mon Mar 13 20:08:25 2017
2017-03-22 14:54:15	index longlist
2017-03-22 15:14:50	index longlist
2017-03-22 15:15:46	index longlist
2017-03-22 15:15:58	index      longlist
2017-03-22 15:25:03	index longlist
2017-03-22 16:28:22	index TCP longlist
2017-03-22 16:30:52	index TCP longlist
2017-03-22 16:33:32	index longlist TCP
2017-03-22 16:34:26	index longlist TCP
2017-03-22 16:38:11	index TCP longlist
2017-03-22 16:38:43	index TCP longlist
2017-03-22 16:39:16	index TCP longlist
2017-03-22 16:39:30	index TCP longlist
2017-03-22 16:40:06	index TCP longlist
2017-03-22 16:41:19	index TCP longlist
2017-03-22 16:43:03	index TCP longlist
2017-03-22 16:44:29	index TCP longlist
2017-03-22 16:46:03	index TCP longlist
2017-03-22 16:46:51	index TCP longlist
2017-03-22 16:47:31	index TCP longlist
2017-03-22 16:48:23	index UDP longlist
2017-03-22 19:42:25	index TCP longlist
2017-03-22 19:47:41	index TCP longlist
2017-03-22 19:48:07	index TCP longlist
2017-03-22 19:48:31	index TCP longlist
2017-03-22 19:49:32	index TCP longlist


201501029/name.py
if protocol == 3:
    logfile.close()
    sys.exit()
if protocol == 4:
    firsttime = 1
    lastsystime = 0
    lasttimedownload = ''
    while (True):
        presentsystime = time.time()

        if (presentsystime - lastsystime > 10):
            presenttime = time.strftime("%c")
            lastsystime = time.time()
            if firsttime:
                res = tcpconnection.send("index longlist", False, ip, directory)
                print res

            else:

                input_raw = "index shortlist " + lasttimedownload + " " + presenttime

                print input_raw
                inputs = input_raw.split(' ')
                messageserver = 'index shortlist ?' + inputs[2] + ' ' + inputs[3] + ' ' + inputs[4] + ' ' + \
                                inputs[5] + ' ' + inputs[6] + ' ? ' + inputs[7] + ' ' + inputs[8] + ' ' + inputs[
                                    9] + ' ' + inputs[10] + ' ' + inputs[11]
                print messageserver

                res = tcpconnection.send(messageserver, False, ip, directory)
                print res
                if res == None:
                    continue
            filelist = []
            word = ''
            linebegin = 1
            for line in res:
                # print line,"completed"


                if linebegin == 1:
                    if line == '\t':
                        linebegin = 0
                        filelist.append(word)
                        word = ''
                    else:
                        word += line
                        print line, word
                        print "camehere"
                if line == '\n':
                    linebegin = 1
                    if not firsttime:
                        filelist.append(word)
                    print word
                    word = ''
            firsttime = 0

            print filelist
            for word in filelist:
                input_raw = "download " + word
                messageserver = 'download ? '
                cnt = 0
                for com in input_raw.split(' '):
                    if cnt >= 1:
                        messageserver += com
                    cnt += 1
                res = udpconnection.send(messageserver, True, ip, directory)
                if res:
                    print res

            lasttimedownload = presenttime


201501113/main.py
#!/bin/python

from threading import Thread
import time
import server
import client
import sys

def start_tcp_server(threadName):
    print 'Started TCP server'
    if len(sys.argv) < 4:
        server.tcp_main(sys.argv[1], 'Share/') #The folder I share with other clients
    else:
        server.tcp_main(sys.argv[1], sys.argv[3]+'/' if sys.argv[3][len(sys.argv[3])-1]!='/' else sys.argv[3])

def start_udp_server(threadName):
    print 'Started UDP server'
    if len(sys.argv) < 4:
        server.udp_main(sys.argv[1], 'Share/') #The folder I share with other clients
    else:
        server.udp_main(sys.argv[1], sys.argv[3]+'/' if sys.argv[3][len(sys.argv[3])-1]!='/' else sys.argv[3])

def start_client(threadName):
    print 'Started client'
    if len(sys.argv) < 4:
        client.main(sys.argv[2], 'ShareDown/') #My downloads get downloaded here
    else:
        client.main(sys.argv[2], sys.argv[3]+'/' if sys.argv[3][len(sys.argv[3])-1]!='/' else sys.argv[3])

t1 = Thread( target=start_tcp_server, args=("Server1", ) )
t1.daemon = True
t1.start()
time.sleep(1)
t2 = Thread( target=start_udp_server, args=("Server2", ) )
t2.daemon = True
t2.start()
time.sleep(1)
t3 = Thread( target=start_client, args=("Client", ) )
t3.start()
       
t3.join()
sys.exit()


201501113/check.py
# client.py

import socket  # Import socket module
import sys
import datetime
import time
import os

logfile = open('history', 'a+')
querycheckfile = open('queryresults', 'a+')
errorlog = open('errors', 'a+')
outputlog = open('outbox', 'a+')


class udp_client:
    def send(self, message, file_request, ip, directory):
        try:
            s = self.connect(ip)
        except Exception, e:
            print str(e)
            errorlog.write(str(e))
            return
        receiving = True
        data = ""
        validity = True
        header = True

        s.sendto(message, (ip, 60001))
        outputlog.write(message)

        if file_request:
            try:
                sequence_number = 0
                file_path = os.path.join(directory, message.split('?')[1].strip())
                if not os.path.exists(os.path.dirname(file_path)):
                    try:
                        os.makedirs(os.path.dirname(file_path))
                        outputlog.write("made a directory with name" + file_path)
                    except Exception, e:
                        print str(e) + ' : Could not create the directory'
                        errorlog.write(str(e) + ' : Could not create the directory')
                        return
                with open(file_path, 'wb') as f:
                    while receiving:
                        data, addr = s.recvfrom(1024)
                        if header and len(data.split('?')) > 4:
                            print 'Received File \n' + data.split('?')[0] + '\n' + data.split('?')[1] + '\n' + \
                                  data.split('?')[2] + '\n' + data.split('?')[3] + '\n' + "End of header"
                            f.write(data.split('?')[4])
                            querycheckfile.write(data.split('?')[4])
                            s.sendto('0', (ip, 60001))
                            outputlog.write('0')
                            header = False
                        elif validity == True:
                            if data.split('#NEXT#')[0] == str(sequence_number + 1):
                                sequence_number += 1
                                s.sendto(str(sequence_number), (ip, 60001))
                                outputlog.write(str(sequence_number))
                                f.write(data.split('#NEXT#')[1])
                                querycheckfile.write(data.split('#NEXT#')[1])
                            else:
                                s.sendto('-1', (ip, 60001))
                                outputlog.write('-1')
                f.close()
                s.close()
                if validity:
                    return "File read"

            except Exception, e:
                print str(e)
                errorlog.write(str(e))
        else:
            try:
                while receiving:
                    data_cur, addr = s.recvfrom(1024)
                    if data_cur == '#END#':
                        receiving = False
                        break
                    data += data_cur
                s.close()
                return data
            except Exception, e:
                print str(e)
                errorlog.write(str(e))

    def connect(self, ip):
        return socket.socket(socket.AF_INET, socket.SOCK_DGRAM)


class tcp_client:
    def connect(self, ip):
        s = socket.socket()
        print "connected to server"
        s.connect((ip, 60000))
        return s

    def send(self, message, file_request, ip, directory):
        try:
            s = self.connect(ip)
            print s
            querycheckfile.write(str(s))
        except Exception, e:
            print str(e) + ' : Failed to connect'
            errorlog.write(str(e) + ' : Failed to connect')
            return
        receiving = True
        data = ""
        validity = True
        header = True

        s.send(message)
        outputlog.write(message)
        if file_request:
            try:
                file_path = os.path.join(directory, message.split('?')[1].strip())
                if not os.path.exists(os.path.dirname(file_path)):
                    try:
                        os.makedirs(os.path.dirname(file_path))
                    except Exception, e:
                        print str(e) + ' : Could not create the directory'
                        errorlog.write(str(e) + ' : Could not create the directory')
                        return
                with open(file_path, 'wb') as f:
                    while True:
                        data = s.recv(1024)
                        if header and len(data.split('?')) > 4:
                            print 'Received File \n' + data.split('?')[0] + '\n' + data.split('?')[1] + '\n' + \
                                  data.split('?')[2] + '\n' + data.split('?')[3] + '\n' + "End of header"
                            f.write(data.split('?')[4])
                            querycheckfile.write(data.split('?')[4])
                            header = False
                        elif validity == True:
                            f.write(data)
                            querycheckfile.write(data.split('?')[4])
                            # write data to a file
                f.close()

                s.close()
                if validity:
                    return "File read"

            except Exception, e:
                print str(e) + ' : Unable to fetch file from server, please enter the correct command'
                errorlog.write(str(e) + ' : Unable to fetch file from server, please enter the correct command')

        else:
            try:
                while receiving:
                    data_cur = s.recv(1024)
                    if not data_cur:
                        receiving = False
                    data += data_cur
                s.close()
                querycheckfile.write(data)
                return data
            except Exception, e:
                print str(e) + ' : Unable to fetch data from server'
                errorlog.write(str(e) + ' : Unable to fetch data from server')




201501113/client.py
import socket  # Import socket module
import sys
import datetime
import time
import os

logfile = open('history', 'a+')
querycheckfile = open('queryresults', 'a+')
errorlog = open('errors', 'a+')
outputlog = open('outbox', 'a+')

class udp_client:
    def send(self, message, file_request, ip, directory):
        try:
            s = self.connect(ip)
        except Exception, e:
            print str(e)
            errorlog.write(str(e))
            return
        receiving = True
        data = ""
        validity = True
        header = True

        s.sendto(message, (ip, 60001))
        outputlog.write(message)

        if file_request:
            try:
                sequence_number = 0
                file_path = os.path.join(directory, message.split('?')[1].strip())
                if not os.path.exists(os.path.dirname(file_path)):
                    try:
                        os.makedirs(os.path.dirname(file_path))
                        outputlog.write("made a directory with name")
                    except Exception, e:
                        print str(e) + ' : Could not create the directory'
                        errorlog.write(str(e) )
                        return
                with open(file_path, 'wb') as f:
                    while receiving:
                        data, addr = s.recvfrom(1024)
                        if data == '#END#':
                            receiving = False
                            break
                        if data == "#101" or data == '#102':
                            validity = False
                        if data == "#102":
                            return 'Please give path to file in shared folder'
                        if not data:
                            break
                        if header and len(data.split('?')) > 4:
                            print 'Received File \n' + data.split('?')[0] + '\n' + data.split('?')[1] + '\n' + \
                                  data.split('?')[2] + '\n' + data.split('?')[3] + '\n' + "End of header"
                            f.write(data.split('?')[4])
                            querycheckfile.write(data.split('?')[4])
                            s.sendto('0', (ip, 60001))
                            outputlog.write('0')
                            header = False
                        elif validity == True:
                            if data.split('#NEXT#')[0] == str(sequence_number + 1):
                                sequence_number += 1
                                s.sendto(str(sequence_number), (ip, 60001))
                                outputlog.write(str(sequence_number))
                                f.write(data.split('#NEXT#')[1])
                                querycheckfile.write(data.split('#NEXT#')[1])
                            else:
                                s.sendto('-1', (ip, 60001))
                                outputlog.write('-1')
                f.close()
                s.close()
                if validity:
                    return "File read"

            except Exception, e:
                print str(e)
                errorlog.write(str(e))
        else:
            try:
                while receiving:
                    data_cur, addr = s.recvfrom(1024)
                    if data_cur == '#END#':
                        receiving = False
                        break
                    data += data_cur
                s.close()
                return data
            except Exception, e:
                print str(e)
                errorlog.write(str(e))

    def connect(self, ip):
        return socket.socket(socket.AF_INET, socket.SOCK_DGRAM)


class tcp_client:
    def connect(self, ip):
        s = socket.socket()
        print "connected to server"
        s.connect((ip, 60006))
        return s

    def send(self, message, file_request, ip, directory):
        try:
            s = self.connect(ip)


        except Exception, e:
            print str(e) + ' : Failed to connect'
            errorlog.write(str(e) + ' : Failed to connect')
            return
        receiving = True
        data = ""
        validity = True
        header = True

        s.send(message)
        outputlog.write(message)

        if file_request:
            try:
                file_path = os.path.join(directory, message.split('?')[1].strip())
                if not os.path.exists(os.path.dirname(file_path)):
                    try:
                        os.makedirs(os.path.dirname(file_path))
                    except Exception, e:
                        print str(e) + ' : Could not create the directory'
                        errorlog.write(str(e) + ' : Could not create the directory')
                        return
                with open(file_path, 'wb') as f:
                    while True:
                        data = s.recv(1024)
                        if data == '#END#':
                            receiving = False
                            break
                        if data == "#101" or data == '#102':
                            validity = False
                        if data == "#102":
                            return 'Please give path to file in shared folder'
                        if not data:
                            break
                        if header and len(data.split('?')) > 4:
                            print 'Received File \n' + data.split('?')[0] + '\n' + data.split('?')[1] + '\n' + \
                                  data.split('?')[2] + '\n' + data.split('?')[3] + '\n' + "End of header"
                            f.write(data.split('?')[4])
                            querycheckfile.write(data.split('?')[4])
                            header = False
                        elif validity == True:
                            f.write(data)
                            querycheckfile.write(data.split('?')[4])
                            # write data to a file
                f.close()

                s.close()
                if validity:
                    return "File read"

            except Exception, e:
                print str(e) + ' : Unable to fetch file from server, please enter the correct command'
                errorlog.write(str(e) + ' : Unable to fetch file from server, please enter the correct command')

        else:
            try:
                while receiving:

                    data_cur = s.recv(1024)
                    if not data_cur:
                        print "2"
                        receiving = False
                    data += data_cur

                s.close()

                querycheckfile.write(data)
                return data
            except Exception, e:
                print str(e) + ' : Unable to fetch data from server'
                errorlog.write(str(e) + ' : Unable to fetch data from server')




def main(ip, directory):
    tcpconnection = tcp_client()
    udpconnection = udp_client()
    type_raw = input("1.general 2.automatic")

    if type_raw == 1:

        while True:
            input_raw = raw_input()
            try:
                if 'index' in input_raw and 'shortlist' in input_raw:
                    logfile.write(datetime.datetime.fromtimestamp(time.time()).strftime(
                        '%Y-%m-%d %H:%M:%S') + '\t' + input_raw + '\n')
                    querycheckfile.write(datetime.datetime.fromtimestamp(time.time()).strftime(
                        '%Y-%m-%d %H:%M:%S') + '\t' + input_raw + '\n')
                    inputs = input_raw.split(' ')
                    messageserver = 'index shortlist ?' + inputs[2] + ' ' + inputs[3] + ' ' + inputs[4] + ' ' + inputs[
                        5] + ' ' + inputs[6] + ' ? ' + inputs[7] + ' ' + inputs[8] + ' ' + inputs[9] + ' ' + inputs[
                                        10] + ' ' + inputs[11]
                    if 'TCP' in input_raw:
                        # IndexGet shortlist Wed Feb 10 15:51:38 2016 Wed Feb 10 15:51:54 2017
                        res = tcpconnection.send(messageserver, False, ip, directory)
                        querycheckfile.write('results:' + res)
                    elif 'UDP' in input_raw:
                        res = udpconnection.send(messageserver, False, ip, directory)
                        querycheckfile.write(('results' + res))
                    if res:
                        print res
                if 'index' in input_raw and 'longlist' in input_raw:
                    logfile.write(datetime.datetime.fromtimestamp(time.time()).strftime(
                        '%Y-%m-%d %H:%M:%S') + '\t' + input_raw + '\n')
                    querycheckfile.write(datetime.datetime.fromtimestamp(time.time()).strftime(
                        '%Y-%m-%d %H:%M:%S') + '\t' + input_raw + '\n')

                    if 'TCP' in input_raw:

                        res = tcpconnection.send("index longlist", False, ip, directory)
                        querycheckfile.write(res)
                    elif 'UDP' in input_raw:
                        res = udpconnection.send("index longlist", False, ip, directory)
                        querycheckfile.write('results' + res)
                    if res:
                        print res

                if 'index' in input_raw and 'regex' in input_raw:
                    logfile.write(datetime.datetime.fromtimestamp(time.time()).strftime(
                        '%Y-%m-%d %H:%M:%S') + '\t' + input_raw + '\n')
                    querycheckfile.write(datetime.datetime.fromtimestamp(time.time()).strftime(
                        '%Y-%m-%d %H:%M:%S') + '\t' + input_raw + '\n')
                    messageserver = 'index regex ?' + input_raw.split(' ')[2]
                    if 'TCP' in input_raw:
                        res = tcpconnection.send(messageserver, False, ip, directory)
                        querycheckfile.write('results' + res)
                    elif 'UDP' in input_raw:
                        res = udpconnection.send(messageserver, False, ip, directory)
                        querycheckfile.write('results' + res)
                    else:
                        res = udpconnection.send(messageserver, False, ip, directory)
                        querycheckfile.write('results' + res)
                    if res:
                        print res

                if 'download' in input_raw:
                    logfile.write(datetime.datetime.fromtimestamp(time.time()).strftime(
                        '%Y-%m-%d %H:%M:%S') + '\t' + input_raw + '\n')
                    querycheckfile.write(input_raw + '\n')
                    messageserver = 'download ? '
                    cnt = 0
                    for com in input_raw.split(' '):
                        if cnt >= 2:
                            messageserver += com
                        cnt += 1
                    if 'TCP' in input_raw:
                        res = udpconnection.send(messageserver, True, ip, directory)
                        querycheckfile.write('results' + str(res))
                    elif 'UDP' in input_raw:
                        res = udpconnection.send(messageserver, True, ip, directory)
                        querycheckfile.write('results' + str(res))
                    if res:
                        print res

                if 'hash' in input_raw:
                    logfile.write(datetime.datetime.fromtimestamp(time.time()).strftime(
                        '%Y-%m-%d %H:%M:%S') + '\t' + input_raw + '\n')
                    querycheckfile.write(datetime.datetime.fromtimestamp(time.time()).strftime(
                        '%Y-%m-%d %H:%M:%S') + '\t' + input_raw + '\n')
                    messageserver = 'hash ' + input_raw.split(' ')[1] + ' ? '
                    cnt = 0
                    for com in input_raw.split(' '):
                        if cnt >= 2:
                            messageserver += com
                        cnt += 1
                    if 'TCP' in input_raw:
                        res = tcpconnection.send(messageserver, False, ip, directory)
                        querycheckfile.write(res)
                    elif 'UDP' in input_raw:
                        res = udpconnection.send(messageserver, False, ip, directory)
                        querycheckfile.write(res)
                    else:
                        res = udpconnection.send(messageserver, False, ip, directory)
                    if res:
                        print res
                    else:
                        print 'Failed to get hash'
            except Exception, e:
                print str(e) + ' : Could not establish connection'
    elif type_raw == 2:

        firsttime = 1
        lastsystime = 0
        lasttimedownload = ''
        while (True):
            presentsystime = time.time()
            #print lastsystime,presentsystime
            if (presentsystime - lastsystime > 10):
                presenttime = time.strftime("%c")
                lastsystime = time.time()
                if firsttime:
                    res = udpconnection.send("index longlist", False, ip, directory)
                    print res

                else:

                    input_raw = "index shortlist " + lasttimedownload + " " + presenttime

                    inputs = input_raw.split(' ')
                    messageserver = 'index shortlist ?' + inputs[2] + ' ' + inputs[3] + ' ' + inputs[4] + ' ' + \
                                    inputs[5] + ' ' + inputs[6] + ' ? ' + inputs[7] + ' ' + inputs[8] + ' ' + inputs[
                                        9] + ' ' + inputs[10] + ' ' + inputs[11]

                    res = udpconnection.send(messageserver, False, ip, directory)
                    if res == None:
                        continue
                filelist = []
                word = ''
                linebegin = 1
                for line in res:
                    # print line,"completed"


                    if linebegin == 1:
                        if line == '\t':
                            linebegin = 0
                            filelist.append(word)
                            word = ''
                        else:
                            word += line
                    if line == '\n':
                        linebegin = 1
                        if not firsttime:
                            filelist.append(word)
                        word = ''
                firsttime = 0

                print filelist
                for word in filelist:
                    input_raw = "download " + word
                    messageserver = 'download ? '
                    cnt = 0
                    for com in input_raw.split(' '):
                        if cnt >= 1:
                            messageserver += com
                        cnt += 1
                    res = udpconnection.send(messageserver, True, ip, directory)
                    if res:
                        print res

                lasttimedownload = presenttime

